File: archive/e/elixir/factorial.ex
defmodule Factorial do
  def main(args) do
    args
    |> List.first
    |> as_integer
    |> get
    |> IO.puts()
  end

  def get(0, acc) do
    acc
  end

  def get(n, acc) do
    get(n - 1, acc * n)
  end

  def get(n) when not is_integer(n) or n < 0 do
    "Usage: please input a non-negative integer"
  end

  def get(n) do
    get(n, 1)
  end

  def as_integer(n) do
    try do
        String.to_integer(n)
    rescue
        ArgumentError -> -1
    end
  end
end

Factorial.main(System.argv())

File: archive/e/erlang/factorial.erl
-module(factorial).
-export([main/1]).

usage() ->
    io:format("Usage: please input a non-negative integer~n"),
    halt().

convert_to_integer(Str) ->
    Result = catch string:to_integer(string:strip(Str)),
    case Result of
        {Int, Rest} when Rest == "" ->
            {ok, Int};
        _ ->
            {error, 0}
    end.

main(Args) ->
    if
        length(Args) < 1 ->
            usage();
        true ->
            ok
    end,

    StrValue = lists:nth(1, Args),
    Value = case convert_to_integer(StrValue) of
        {ok, Int} ->
            Int;
        _ ->
            usage()
    end,

    if
        Value < 0 ->
            usage();
        true ->
            ok
    end,

    FValue = factorial(Value),
    io:format("~w~n", [FValue]).

%%--------------------------------------------------------------------
%% Recursively multiply N times N-1 until N <= 1
%%--------------------------------------------------------------------
factorial(0) ->
    1;
factorial(N) ->
    N * factorial(N-1).
